README.md
# Experimental Stuff with Haskell

This repository contains some experimental stuff while learning fundamentals of Haskell.


ChangeLog.md
# Revision history for black-jack

## 0.1.0.0  -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


test/BlackJackSpec.hs
module BlackJackSpec (spec) where

import BlackJack
import Test.Hspec
import Test.Hspec.QuickCheck

spec :: Spec
spec = do
    describe "dealer" $ do
        it "stands on hard seventeen" $ do
            getCards (dealerAction [] (Hand [Card Ten Club, Card Seven Spade] False False)) 
                `shouldBe` ([Card Ten Club, Card Seven Spade])

        it "hits on hard sixteen" $ do
            getCards (dealerAction [Card Eight Club] (Hand [Card Ten Club, Card Six Spade] False False)) 
                `shouldBe` ([Card Ten Club, Card Six Spade, Card Eight Club])

        it "stands on soft seventeen" $ do
            getCards (dealerAction [] (Hand [Card Ace Club, Card Six Spade] False False)) 
                `shouldBe` ([Card Ace Club, Card Six Spade])

        it "hits on soft sixteen" $ do
            getCards (dealerAction [Card Four Club] (Hand [Card Ace Club, Card Five Spade] False False)) 
                `shouldBe` ([Card Ace Club, Card Five Spade, Card Four Club])

        it "hits on soft bust" $ do
            getCards (dealerAction [Card Six Club, Card Two Heart, Card Three Diamond] (Hand [Card Ace Club, Card Five Spade] False False)) 
                `shouldBe` ([Card Ace Club, Card Five Spade, Card Six Club, Card Two Heart, Card Three Diamond])

black-jack.hs
import Data.List
import Control.Monad
import System.Random

-- ## Type definition ##

data RoundResult = PlayerWin | PlayerBust | StandOff | PlayerLoose | PlayerBlackJack | Undefined
    deriving (Read, Show, Enum, Eq, Ord)

data Suit = Club | Diamond | Heart | Spade
    deriving (Read, Show, Enum, Eq, Ord)

data CardValue = Two | Three | Four | Five 
    | Six | Seven | Eight | Nine | Ten 
    | Jack | Queen | King | Ace
    deriving (Read, Show, Enum, Eq, Ord)

data Card = Card CardValue Suit
    deriving (Read, Show, Eq)

-- provide instance declarations for Ord and Enum
instance Ord Card where
    compare c1 c2 = compare (getCardValue c1, getCardSuit c1) (getCardValue c2, getCardSuit c2)

instance Enum Card where
    toEnum n = let (v, s) = n `divMod` 4 in Card (toEnum v) (toEnum s)
    fromEnum c = fromEnum (getCardValue c) * 4 + fromEnum (getCardSuit c)

-- |Declare 'Deck' as a synonym of array of 'Card'
type Deck = [Card]

type Hand = [Card]

data PlayerDeck = PlayerDeck Hand Float
    deriving (Show)

-- |Declare 'Box' as a synonym of array of 'Deck'
type Box = [PlayerDeck]

data Table = Table Deck Hand Box
    deriving (Show)

hardCardValue :: Card -> Int
hardCardValue c
    | getCardValue c == Two   = 2
    | getCardValue c == Three = 3
    | getCardValue c == Four  = 4
    | getCardValue c == Five  = 5
    | getCardValue c == Six   = 6
    | getCardValue c == Seven = 7
    | getCardValue c == Eight = 8
    | getCardValue c == Nine  = 9
    | getCardValue c == Ten   = 10
    | getCardValue c == Jack  = 10
    | getCardValue c == Queen = 10
    | getCardValue c == King  = 10
    | otherwise               = 0

-- |Returns the Deck of a snapshot
getDeck :: Table -> Deck
getDeck (Table deck _ _) = deck

-- |Returns the Dealer Hand of the table
getDealerHand :: Table -> Hand
getDealerHand (Table _ dealerHand _) = dealerHand

getPlayerHand :: PlayerDeck -> Hand
getPlayerHand (PlayerDeck hand _) = hand

getHandBet :: PlayerDeck -> Float
getHandBet (PlayerDeck _ bet) = bet

-- |Returns the Player Decks of a snapshot
getPlayerBox :: Table -> Box
getPlayerBox (Table _ _ playerBox) = playerBox

-- |Returns 'True' if the Hand is Black Jack
isBlackJack :: Hand -> Bool
isBlackJack hand = length hand == 2 && (any (\c -> getCardValue c == Ace) hand 
    && ((any (\c -> getCardValue c == Ten) hand)
        || any (\c -> getCardValue c == Jack) hand
        || any (\c -> getCardValue c == Queen) hand
        || any (\c -> getCardValue c == King) hand))

isNotBlackJack :: Hand -> Bool
isNotBlackJack hand = not (isBlackJack hand)

-- |Returns the Value of a card
getCardValue :: Card -> CardValue
getCardValue (Card val _) = val

-- |Returns the Suit of a card
getCardSuit :: Card -> Suit
getCardSuit (Card _ suit) = suit

-- |Returns 'True' if the hand value is > 21
isBust :: Hand -> Bool
isBust hand = (handValue hand) > 21

isNotBust :: Hand -> Bool
isNotBust hand = not (isBust hand)

-- |Returns the value of the hand under the assumption that the number of hard aces must be minimal
handValue :: Hand -> Int
handValue hand = sumCardValues hand (acesAsOneCount hand)

-- |Returns the sum of the hand by counting the specified number of aces as one
sumCardValues :: Hand -> Int -> Int
sumCardValues hand acesAsOneCount =
    (sum $ map hardCardValue (filterNonAces hand)) + 
    (1 * acesAsOneCount) + 
    (11 * ((aceCount hand) - acesAsOneCount))

-- |Returns the number of Aces in the Deck
aceCount :: Hand -> Int
aceCount hand = length (filter isAce hand)

-- |Returns 'True' if the card is an Ace
isAce :: Card -> Bool
isAce card = getCardValue card == Ace

-- |Returns a filtered Deck without any aces
filterNonAces :: Hand -> Hand
filterNonAces hand = filter (not . isAce) hand

-- |Returns the number of Aces to be counted as one in the dealer hand so that the hand value is <=21
acesAsOneCount :: Hand -> Int
acesAsOneCount hand = acesAsOneCountLoop hand 0

-- |Recursive loop to determine the minimum number of Aces to be counted as 1 so that the hand value is <=21
acesAsOneCountLoop :: Hand -> Int -> Int
acesAsOneCountLoop hand n = 
    if n < aceCount hand && sumCardValues hand n > 21
    then acesAsOneCountLoop hand (n+1)
    else n

boxCardCount :: Box -> Int
boxCardCount box = foldl (+) 0 (map (\d -> length (getPlayerHand d)) box)

boxHandCount :: Box -> Int
boxHandCount box = length box

-- |Returns a shuffled deck with the specified number of 52-card decks
staticShuffledDeck :: Int -> Deck
staticShuffledDeck deckCount = shuffle (createDeck deckCount)

-- |Shuffles a deck with up to 312 cards
shuffle :: Deck -> Deck
shuffle deck =
    map snd (sort (zip (filter (\r -> r < length deck) rand) deck))

-- |Returns randomized list of numbers [0..311] (6 decks à 52 cards)
rand :: [Int]
rand = [287, 202, 306, 170, 15, 125, 104, 261, 117, 305, 255, 301, 246, 78, 239, 79, 98, 281, 296, 45, 210, 273, 307, 227, 41, 10, 50, 149, 252, 8, 107, 11, 53, 217, 97, 90, 247, 74, 309, 212, 184, 95, 48, 153, 81, 262, 83, 203, 272, 63, 264, 75, 187, 302, 145, 110, 300, 31, 57, 185, 249, 55, 282, 199, 100, 173, 228, 230, 24, 267, 84, 128, 62, 148, 191, 303, 133, 297, 0, 205, 198, 193, 77, 279, 36, 269, 17, 118, 85, 152, 111, 89, 28, 38, 190, 208, 216, 22, 142, 87, 113, 58, 167, 169, 288, 52, 68, 26, 195, 124, 201, 204, 70, 260, 115, 280, 292, 139, 138, 233, 5, 134, 144, 64, 225, 130, 14, 49, 164, 241, 200, 259, 285, 243, 215, 213, 82, 23, 189, 33, 30, 291, 32, 235, 155, 207, 34, 265, 127, 294, 234, 289, 299, 276, 156, 157, 37, 242, 120, 238, 222, 275, 286, 76, 9, 165, 46, 160, 310, 158, 71, 27, 122, 268, 250, 209, 119, 88, 67, 180, 103, 132, 25, 54, 197, 93, 7, 244, 168, 3, 61, 99, 214, 283, 56, 256, 179, 35, 166, 143, 254, 263, 290, 129, 51, 220, 226, 60, 20, 223, 298, 101, 80, 181, 177, 12, 172, 91, 39, 271, 311, 1, 206, 140, 163, 69, 96, 43, 29, 266, 194, 150, 109, 131, 182, 102, 188, 284, 59, 295, 231, 47, 232, 42, 137, 221, 248, 65, 253, 154, 274, 293, 192, 66, 304, 105, 178, 147, 240, 114, 73, 308, 116, 13, 161, 162, 196, 258, 270, 245, 171, 106, 175, 218, 174, 224, 278, 19, 211, 251, 92, 135, 44, 146, 141, 40, 236, 2, 6, 237, 186, 126, 121, 123, 229, 21, 16, 136, 159, 94, 112, 4, 18, 86, 183, 151, 219, 72, 176, 257, 108, 277]

-- |Returns the specified number of 52-card decks as on ordered list
createDeck :: Int -> Deck
createDeck deckCount =
    if deckCount <= 0 || deckCount > 6
    then []
    else [Card val su | val <- [Two .. Ace], su <- [Club .. Spade]] ++ createDeck (deckCount - 1)

bet :: Float
bet = 1.0

-- |Returns the number of rounds to play
numberOfRounds :: Int
numberOfRounds = 10

numberOfDecks :: Int
numberOfDecks = 6

play :: IO ()
play = do 
    let deck = (staticShuffledDeck numberOfDecks)
    putStrLn (show (take 21 (drop 28 deck)))
    playLoop deck numberOfRounds 0 0

playLoop :: Deck -> Int -> Int -> Float -> IO ()
playLoop deck numberOfRounds currentRound balance = do 
    if (currentRound <= numberOfRounds)
    then do 
        putStrLn ("Playing loop (" ++ show currentRound ++ "/" ++ show numberOfRounds ++ ") with " ++ show (length deck) ++ " cards (" ++ show ((numberOfDecks * 52) - length deck) ++ " cards played).")

        let table = playBox (drop 3 deck) [(head (drop 1 deck))] (PlayerDeck [head deck, head (drop 2 deck)] bet)
        let earnings = evaluateEarnings table

        putStrLn ("  > Dealer Deck (" ++ (if isBlackJack (getDealerHand table) then "BJ" else show (handValue (getDealerHand table))) ++ "): " ++ show (getDealerHand table))
        mapM (\ d -> putStrLn ("  > Player Deck (" ++ (if (isBlackJack (getPlayerHand d)) then "BJ" else show (handValue (getPlayerHand d))) ++ "): " ++ show (getPlayerHand d))) (getPlayerBox table)
        putStrLn ("  > Yield: " ++ (if earnings > 0 then "+" else "") ++ show earnings)
        putStrLn ("  > Balance: " ++ show (balance + earnings))
        playLoop (getDeck table) numberOfRounds (currentRound + 1) (balance + earnings)
    else putStrLn ("End of Game")

playBox :: Deck -> Hand -> PlayerDeck -> Table
playBox deck dealerHand playerDeck = do
    let playerBox = playHand deck dealerHand playerDeck
    let playerBoxDealedCardCount = boxCardCount playerBox - 2

    let finalDealerHand = dealerAction (drop (playerBoxDealedCardCount) deck) dealerHand
    let dealerHandDealedCardCount = length finalDealerHand - 1
    
    Table (drop (playerBoxDealedCardCount + dealerHandDealedCardCount) deck) finalDealerHand playerBox

dealerAction :: Deck -> Hand -> Hand
dealerAction deck dealerHand 
    | handValue dealerHand < 17 = dealerAction deck (dealerHand ++ [head deck])
    | otherwise = dealerHand

playHand :: Deck -> Hand -> PlayerDeck -> Box
playHand deck dealerHand playerDeck 
    | canSplit (getPlayerHand playerDeck) = do
        let left = playHand deck dealerHand (PlayerDeck [head (getPlayerHand playerDeck)] (getHandBet playerDeck))
        let right = playHand (drop (boxCardCount left - 1) deck) dealerHand (PlayerDeck [head (tail (getPlayerHand playerDeck))] (getHandBet playerDeck))
        left ++ right
    | canHit (getPlayerHand playerDeck) && voteHit dealerHand (getPlayerHand playerDeck) = foldl (++) [] (map (\d -> playHand (tail deck) dealerHand d) [PlayerDeck ((getPlayerHand playerDeck) ++ [head deck]) (getHandBet playerDeck)])
    | otherwise = [playerDeck]

canSplit :: Hand -> Bool
canSplit hand = length hand == 2 && handValue (take 1 hand) == handValue (drop 1 hand)

voteSplit :: Hand -> Hand -> Bool
voteSplit dealerHand playerHand = True

canHit :: Hand -> Bool
canHit hand = handValue hand <= 21

voteHit :: Hand -> Hand -> Bool
voteHit dealerHand playerHand = handValue playerHand < 17

evaluateEarnings :: Table -> Float
evaluateEarnings table =
    winHandSum table
    + blackJackHandSum table
    - looseHandSum table
    - bustHandSum table

isDealerBust :: Table -> Bool
isDealerBust table = isBust (getDealerHand table)

isDealerNotBust :: Table -> Bool
isDealerNotBust table = not (isDealerBust table)

bustHandSum :: Table -> Float
bustHandSum table = foldl (+) 0 (map (\h -> (getHandBet h))
    (filter (\d -> isBust (getPlayerHand d)) (getPlayerBox table)))

looseHandSum :: Table -> Float
looseHandSum table = foldl (+) 0 (map (\h -> (getHandBet h))
    (filter (\d -> (isNotBlackJack (getPlayerHand d)
    && isNotBlackJack (getDealerHand table)
    && isNotBust (getPlayerHand d)
    && isDealerNotBust table 
    && (handValue (getPlayerHand d) < handValue (getDealerHand table)))
    || (isNotBlackJack (getPlayerHand d) 
        && isBlackJack (getDealerHand table))) (getPlayerBox table)))

winHandSum :: Table -> Float
winHandSum table = foldl (+) 0 (map (\h -> (getHandBet h)) 
    (filter (\d -> isNotBlackJack (getPlayerHand d)
    && isNotBlackJack (getDealerHand table)
    && isNotBust (getPlayerHand d)
    && isDealerNotBust table 
    && (handValue (getPlayerHand d) > handValue (getDealerHand table))) (getPlayerBox table)))

blackJackHandSum :: Table -> Float
blackJackHandSum table = foldl (+) 0 (map (\h -> 3.0 / 2.0 * (getHandBet h))
    (filter (\d -> isBlackJack (getPlayerHand d) 
    && isNotBlackJack (getDealerHand table)) (getPlayerBox table)))

src/exe/Main.hs
module Main (main) where

import BlackJack
import Data.List

-- |Returns randomized list of numbers [0..311] (6 decks à 52 cards)
rand :: [Int]
rand = [287, 202, 306, 170, 15, 125, 104, 261, 117, 305, 255, 301, 246, 78, 239, 79, 98, 281, 296, 45, 210, 273, 307, 227, 41, 10, 50, 149, 252, 8, 107, 11, 53, 217, 97, 90, 247, 74, 309, 212, 184, 95, 48, 153, 81, 262, 83, 203, 272, 63, 264, 75, 187, 302, 145, 110, 300, 31, 57, 185, 249, 55, 282, 199, 100, 173, 228, 230, 24, 267, 84, 128, 62, 148, 191, 303, 133, 297, 0, 205, 198, 193, 77, 279, 36, 269, 17, 118, 85, 152, 111, 89, 28, 38, 190, 208, 216, 22, 142, 87, 113, 58, 167, 169, 288, 52, 68, 26, 195, 124, 201, 204, 70, 260, 115, 280, 292, 139, 138, 233, 5, 134, 144, 64, 225, 130, 14, 49, 164, 241, 200, 259, 285, 243, 215, 213, 82, 23, 189, 33, 30, 291, 32, 235, 155, 207, 34, 265, 127, 294, 234, 289, 299, 276, 156, 157, 37, 242, 120, 238, 222, 275, 286, 76, 9, 165, 46, 160, 310, 158, 71, 27, 122, 268, 250, 209, 119, 88, 67, 180, 103, 132, 25, 54, 197, 93, 7, 244, 168, 3, 61, 99, 214, 283, 56, 256, 179, 35, 166, 143, 254, 263, 290, 129, 51, 220, 226, 60, 20, 223, 298, 101, 80, 181, 177, 12, 172, 91, 39, 271, 311, 1, 206, 140, 163, 69, 96, 43, 29, 266, 194, 150, 109, 131, 182, 102, 188, 284, 59, 295, 231, 47, 232, 42, 137, 221, 248, 65, 253, 154, 274, 293, 192, 66, 304, 105, 178, 147, 240, 114, 73, 308, 116, 13, 161, 162, 196, 258, 270, 245, 171, 106, 175, 218, 174, 224, 278, 19, 211, 251, 92, 135, 44, 146, 141, 40, 236, 2, 6, 237, 186, 126, 121, 123, 229, 21, 16, 136, 159, 94, 112, 4, 18, 86, 183, 151, 219, 72, 176, 257, 108, 277]

-- |Shuffles a deck with up to 312 cards
shuffle :: Deck -> Deck
shuffle deck =
    map snd (sort (zip (filter (\r -> r < length deck) rand) deck))

-- |Returns a shuffled deck with the specified number of 52-card decks
staticShuffledDeck :: Int -> Deck
staticShuffledDeck deckCount = shuffle (createDeck deckCount)

numberOfDecks :: Int
numberOfDecks = 6

-- | Returns the number of rounds to play
initNumberOfRounds :: Int
initNumberOfRounds = 40

main :: IO ()
main = do 
    putStrLn "Lets play some Black Jack!"
    -- let deck = [Card Nine Spade, Card Five Heart, Card Nine Diamond, Card Two Heart, Card Five Club, Card Ten Heart, Card King Spade, Card Five Diamond]
    let deck = staticShuffledDeck numberOfDecks
    let tables = play deck initNumberOfRounds
    putStrLn (showTables tables)
    putStrLn "Game ended."


src/lib/BlackJack.hs
module BlackJack (play, getCards, Table, Deck, Hand(..), dealerAction, createDeck, showTables, Card(..), CardValue (Two, Three, Four, Five, Six, Seven, Eight, Nine, Ten, Jack, Queen, King, Ace), Suit (Spade, Diamond, Club, Heart)) where

data RoundResult = PlayerWin | PlayerBust | StandOff | PlayerLoose | PlayerBlackJack | Undefined
    deriving (Read, Show, Enum, Eq, Ord)

data Suit = Club | Diamond | Heart | Spade
    deriving (Read, Show, Enum, Eq, Ord)

data CardValue = Two | Three | Four | Five 
    | Six | Seven | Eight | Nine | Ten 
    | Jack | Queen | King | Ace
    deriving (Read, Show, Enum, Eq, Ord)

data Card = Card CardValue Suit
    deriving (Eq)

instance Show Card where
    show c = show (getCardValue c) ++ " of " ++ show (getCardSuit c)

-- provide instance declarations for Ord and Enum
instance Ord Card where
    compare c1 c2 = compare (getCardValue c1, getCardSuit c1) (getCardValue c2, getCardSuit c2)

instance Enum Card where
    toEnum n = let (v, s) = n `divMod` 4 in Card (toEnum v) (toEnum s)
    fromEnum c = fromEnum (getCardValue c) * 4 + fromEnum (getCardSuit c)

-- |Declare 'Deck' as a synonym of array of 'Card'
type Deck = [Card]

data Hand = Hand [Card] Bool Bool
    deriving (Show)

data PlayerDeck = PlayerDeck Hand Float
    deriving (Show)

-- | Declare 'Box' as a synonym of array of 'Deck'
data Box = Box [PlayerDeck] Int
    deriving (Show)

data Table = Table Deck Hand Box Int
    deriving (Show)

shortCardValue :: CardValue -> String
shortCardValue value 
    | value == Two = "2"
    | value == Three = "3"
    | value == Four = "4"
    | value == Five = "5"
    | value == Six = "6"
    | value == Seven = "7"
    | value == Eight = "8"
    | value == Nine = "9"
    | value == Ten = "10"
    | value == Jack = "J"
    | value == Queen = "Q"
    | value == King = "K"
    | value == Ace = "A"
    | otherwise = "x"

shortCardSuit :: Suit -> String
shortCardSuit suit 
    | suit == Spade = "S"
    | suit == Club = "C"
    | suit == Diamond = "D"
    | suit == Heart = "H"
    | otherwise = "x"

-- |Returns the specified number of 52-card decks as on ordered list
createDeck :: Int -> Deck
createDeck deckCount =
    if deckCount <= 0 || deckCount > 6
    then []
    else [Card val su | val <- [Two .. Ace], su <- [Club .. Spade]] ++ createDeck (deckCount - 1)

shortCard :: Card -> String
shortCard card = shortCardValue (getCardValue card) ++ shortCardSuit (getCardSuit card)

showTables :: [Table] -> String
showTables tables = concatMap showTable tables ++
    "    > Balance: " ++
        show (foldl (\ x y -> x + evaluateEarnings y) 0 tables)

showTable :: Table -> String
showTable table = "Playing loop " ++ show (getLoopNumber table) ++ "\r\n" ++ showDealerDeck table ++ "\r\n" ++ showPlayerBox table ++ "\r\n" ++ "    > Preview: " ++ showDeckPreview table ++ "\r\n"

showDeckPreview :: Table -> String
showDeckPreview table = show (take 15 (map shortCard (getDeck table)))

showDealerDeck :: Table -> String
showDealerDeck table = "    > Dealer Deck (" ++ (if isBlackJack (getDealerHand table) then "BJ" else show (handValue (getDealerHand table))) ++ "): " ++ show (getCards (getDealerHand table))

showPlayerBox :: Table -> String
showPlayerBox table = "    > Box SplitCount: " ++ show (getSplitCount (getPlayerBox table)) ++ "\r\n" ++
    concatMap showPlayerDeck (getPlayerDecks (getPlayerBox table)) 
        ++ "        > Yield: " ++ (if evaluateEarnings table > 0 then "+" else "") ++ show (evaluateEarnings table)

showPlayerDeck :: PlayerDeck -> String
showPlayerDeck playerDeck = "        > Player Deck [" ++ showHandMode (getPlayerHand playerDeck) ++ "] (" ++
        (if isBlackJack (getPlayerHand playerDeck) then "BJ" else show (handValue (getPlayerHand playerDeck)))
            ++ "): " ++ show (getCards (getPlayerHand playerDeck)) ++ "\r\n"

showHandMode :: Hand -> String
showHandMode hand = (if isSplittedHand hand then "S" else "-") ++ (if isDoubled hand then "D" else "-")


hardCardValue :: Card -> Int
hardCardValue c
    | getCardValue c == Two   = 2
    | getCardValue c == Three = 3
    | getCardValue c == Four  = 4
    | getCardValue c == Five  = 5
    | getCardValue c == Six   = 6
    | getCardValue c == Seven = 7
    | getCardValue c == Eight = 8
    | getCardValue c == Nine  = 9
    | getCardValue c == Ten   = 10
    | getCardValue c == Jack  = 10
    | getCardValue c == Queen = 10
    | getCardValue c == King  = 10
    | otherwise               = 0

getCardCount :: Table -> Int
getCardCount table = length (getCards (getDealerHand table)) + foldl (\x y -> x + length (getCards (getPlayerHand y))) 0 (getPlayerDecks (getPlayerBox table))

getLoopNumber :: Table -> Int
getLoopNumber (Table _ _ _ l) = l

-- |Returns the Deck of a snapshot
getDeck :: Table -> Deck
getDeck (Table deck _ _ _) = deck

-- |Returns the Dealer Hand of the table
getDealerHand :: Table -> Hand
getDealerHand (Table _ dealerHand _ _) = dealerHand

getPlayerHand :: PlayerDeck -> Hand
getPlayerHand (PlayerDeck hand _ ) = hand

getBet :: PlayerDeck -> Float
getBet (PlayerDeck _ b) = b

getCards :: Hand -> [Card]
getCards (Hand cards _ _) = cards

getSplitCount :: Box -> Int
getSplitCount (Box _ c) = c

getPlayerDecks :: Box -> [PlayerDeck]
getPlayerDecks (Box p _) = p

isSplitted :: PlayerDeck -> Bool
isSplitted playerDeck = isSplittedHand (getPlayerHand playerDeck)

isSplittedHand :: Hand -> Bool
isSplittedHand (Hand _ split _) = split

isDoubled :: Hand -> Bool
isDoubled (Hand _ _ double) = double

-- |Returns the Player Decks of a snapshot
getPlayerBox :: Table -> Box
getPlayerBox (Table _ _ playerBox _) = playerBox

-- |Returns 'True' if the Hand is Black Jack
isBlackJack :: Hand -> Bool
isBlackJack hand = not (isSplittedHand hand) && 
    length (getCards hand) == 2 && (any (\c -> getCardValue c == Ace) (getCards hand)
        && (any (\ c -> getCardValue c == Ten) (getCards hand) ||
            any (\ c -> getCardValue c == Jack) (getCards hand) ||
                any (\ c -> getCardValue c == Queen) (getCards hand) ||
                    any (\ c -> getCardValue c == King) (getCards hand)))

isNotBlackJack :: Hand -> Bool
isNotBlackJack hand = not (isBlackJack hand)

-- |Returns the Value of a card
getCardValue :: Card -> CardValue
getCardValue (Card val _) = val

-- |Returns the Suit of a card
getCardSuit :: Card -> Suit
getCardSuit (Card _ suit) = suit

-- |Returns 'True' if the hand value is > 21
isBust :: Hand -> Bool
isBust hand = handValue hand > 21

isNotBust :: Hand -> Bool
isNotBust hand = not (isBust hand)

handValue :: Hand -> Int
handValue hand = if isDoubled hand then sumCardValues (getCards hand) 0 else sumCardValues (getCards hand) (acesAsOneCount (getCards hand))

-- |Returns the sum of the hand by counting the specified number of aces as one
sumCardValues :: [Card] -> Int -> Int
sumCardValues cards acesOsAne =
    sum (map hardCardValue (filterNonAces cards)) +
        (1 * acesOsAne) + 
    (11 * (aceCount cards - acesOsAne))

-- |Returns the number of Aces in the Deck
aceCount :: [Card] -> Int
aceCount cards = length (filter isAce cards)

-- |Returns 'True' if the card is an Ace
isAce :: Card -> Bool
isAce card = getCardValue card == Ace

-- |Returns a filtered Deck without any aces
filterNonAces :: [Card] -> [Card]
filterNonAces = filter (not . isAce)

-- |Returns the number of Aces to be counted as one in the dealer hand so that the hand value is <=21
acesAsOneCount :: [Card] -> Int
acesAsOneCount cards = acesAsOneCountLoop cards 0

-- |Recursive loop to determine the minimum number of Aces to be counted as 1 so that the hand value is <=21
acesAsOneCountLoop :: [Card] -> Int -> Int
acesAsOneCountLoop cards n = 
    if n < aceCount cards && sumCardValues cards n > 21
    then acesAsOneCountLoop cards (n+1)
    else n

boxCardCount :: Box -> Int
boxCardCount box = sum (map (length . getCards . getPlayerHand) (getPlayerDecks box))

bet :: Float
bet = 1.0

play :: Deck -> Int -> [Table]
play deck numberOfRounds = do 
    playLoop deck numberOfRounds 0 0

playLoop :: Deck -> Int -> Int -> Float -> [Table]
playLoop deck numberOfRounds roundsPlayed balance = do
    let table = (playBox (drop 3 deck) (Hand [deck !! 1] False False) (PlayerDeck (Hand [head deck, deck !! 2] False False) bet)) roundsPlayed
    let earnings = evaluateEarnings table
    if roundsPlayed < numberOfRounds 
        then table : playLoop (getDeck table) numberOfRounds (roundsPlayed + 1) (balance + earnings)
        else [table]

playBox :: Deck -> Hand -> PlayerDeck -> Int -> Table
playBox deck dealerHand playerDeck loopNumber = do
    let playerBox = playHand deck 0 dealerHand playerDeck
    let playerBoxDealedCardCount = boxCardCount playerBox - 2

    let finalDealerHand = dealerAction (drop playerBoxDealedCardCount deck) dealerHand
    let dealerHandDealedCardCount = length (getCards finalDealerHand) - 1
    
    Table (drop (playerBoxDealedCardCount + dealerHandDealedCardCount) deck) finalDealerHand playerBox loopNumber

dealerAction :: Deck -> Hand -> Hand
dealerAction deck dealerHand 
    | handValue dealerHand < 17 = dealerAction (tail deck) (Hand (getCards dealerHand ++ [head deck]) False False)
    | otherwise = dealerHand

playHand :: Deck -> Int -> Hand -> PlayerDeck -> Box
playHand deck splitCount dealerHand playerDeck
    | canSplit (getPlayerHand playerDeck) splitCount && voteSplit dealerHand (getPlayerHand playerDeck) = do
        let leftCard = head (getCards (getPlayerHand playerDeck))
        let rightCard = head (tail (getCards (getPlayerHand playerDeck)))
        let left = if isAce leftCard
            then Box [PlayerDeck (Hand (getCards (getPlayerHand playerDeck) ++ [head deck]) True False) (getBet playerDeck)] (splitCount + 1)
            else playHand deck (splitCount + 1) dealerHand (PlayerDeck (Hand [leftCard] True False) (getBet playerDeck))
        let right = if isAce rightCard 
            then Box [PlayerDeck (Hand (getCards (getPlayerHand playerDeck) ++ [head (drop (boxCardCount left - 1) deck)]) True False) (getBet playerDeck)] (splitCount + 1)
            else playHand (drop (boxCardCount left - 1) deck) (getSplitCount left) dealerHand (PlayerDeck (Hand [rightCard] True False) (getBet playerDeck))
        Box (getPlayerDecks left ++ getPlayerDecks right) (getSplitCount right)
    | canDouble (getPlayerHand playerDeck) && voteDouble (getPlayerHand playerDeck) = 
        Box [PlayerDeck (Hand (getCards (getPlayerHand playerDeck) ++ [head deck]) (isSplitted playerDeck) True) (2.0 * getBet playerDeck)] splitCount
    | canHit (getPlayerHand playerDeck) && voteHit dealerHand (getPlayerHand playerDeck) = 
        foldl (\x y -> Box (getPlayerDecks x ++ getPlayerDecks y) (getSplitCount y)) (Box [] 0) 
            (map (playHand (tail deck) splitCount dealerHand) [PlayerDeck (Hand (getCards (getPlayerHand playerDeck) ++ [head deck]) (isSplitted playerDeck) False) (getBet playerDeck)])
    | otherwise = Box [playerDeck] splitCount

-- |Returns 'True' if the two cards have the same value
canSplit :: Hand -> Int -> Bool
canSplit hand splitCount = 
    splitCount < 3 
    && length (getCards hand) == 2 
    && handValue (Hand (take 1 (getCards hand)) (isSplittedHand hand) (isDoubled hand)) 
        == handValue (Hand (drop 1 (getCards hand)) (isSplittedHand hand) (isDoubled hand))

voteSplit :: Hand -> Hand -> Bool
voteSplit dealerHand playerHand = True

-- |Returns 'True' if the value of the hand is less than 21 and the hand is not doubled
canHit :: Hand -> Bool
canHit hand = handValue hand < 21 && not (isDoubled hand)

voteHit :: Hand -> Hand -> Bool
voteHit dealerHand playerHand = handValue playerHand < 17

-- |Returns 'True' if the two cards of the hand sum up to 9, 10 or 11
canDouble :: Hand -> Bool
canDouble hand = length (getCards hand) == 2 -- && (handValue hand == 9 || handValue hand == 10 || handValue hand == 11)

voteDouble :: Hand -> Bool
voteDouble hand = True

evaluateEarnings :: Table -> Float
evaluateEarnings table =
    winHandSum table
    + blackJackHandSum table
    - looseHandSum table
    - bustHandSum table

isDealerBust :: Table -> Bool
isDealerBust table = isBust (getDealerHand table)

isDealerNotBust :: Table -> Bool
isDealerNotBust table = not (isDealerBust table)

bustHandSum :: Table -> Float
bustHandSum table = sum (map getBet
    (filter (isBust . getPlayerHand) (getPlayerDecks (getPlayerBox table))))

looseHandSum :: Table -> Float
looseHandSum table = sum (map getBet
    (filter (\d -> (isNotBlackJack (getPlayerHand d)
    && isNotBlackJack (getDealerHand table)
    && isNotBust (getPlayerHand d)
    && isDealerNotBust table 
    && (handValue (getPlayerHand d) < handValue (getDealerHand table)))
    || (isNotBlackJack (getPlayerHand d) 
        && isBlackJack (getDealerHand table))) (getPlayerDecks (getPlayerBox table))))

winHandSum :: Table -> Float
winHandSum table = sum (map getBet 
    (filter (\d -> isNotBlackJack (getPlayerHand d)
    && isNotBlackJack (getDealerHand table)
    && isNotBust (getPlayerHand d)
    && isDealerNotBust table 
    && (handValue (getPlayerHand d) > handValue (getDealerHand table))) (getPlayerDecks (getPlayerBox table))))

blackJackHandSum :: Table -> Float
blackJackHandSum table = sum (map (\h -> 3.0 / 2.0 * getBet h)
    (filter (\d -> isBlackJack (getPlayerHand d) 
    && isNotBlackJack (getDealerHand table)) (getPlayerDecks (getPlayerBox table))))
